src/librustc_typeck/collect.rs

```diff
@@ -1605,27 +1611,24 @@ fn early_bound_lifetimes_from_generics<'a, 'tcx>(
 fn predicates_defined_on<'a, 'tcx>(
     tcx: TyCtxt<'a, 'tcx, 'tcx>,
     def_id: DefId,
-) -> ty::GenericPredicates<'tcx> {
+) -> Lrc<ty::GenericPredicates<'tcx>> {
     let explicit = tcx.explicit_predicates_of(def_id);
     let span = tcx.def_span(def_id);
-    let predicates = explicit.predicates.into_iter().chain(
+    let predicates = explicit.predicates.iter().cloned().chain(
```

Comment:
A lot of time, `inferred_outlives_of` will be empty -- maybe we should check for that?

If we added `Lrc::make_mut`, we could do:

```rust
let mut predicates = tcx.explicit_predicates_of(def_id);
let inferred_outlives = tcx.inferred_outlives_of(def_id);
if !inferred_outlives.is_empty() {
    let predicates = Lrc::make_mut(&mut predicates);
    predicates.predicates.extend(inferred_outlives.iter().map(|&p| (p, span)));
}
```

or something like that